pkg/kube/krt/static.go

```diff
@@ -15,29 +15,96 @@
 package krt
 
 import (
+	"sync"
+
 	"istio.io/istio/pkg/kube/controllers"
 	"istio.io/istio/pkg/kube/kclient"
 	"istio.io/istio/pkg/maps"
 	"istio.io/istio/pkg/slices"
 )
 
+type StaticCollection[T any] struct {
+	*staticList[T]
+}
+
 type staticList[T any] struct {
-	vals map[Key[T]]T
-	id   collectionUID
+	mu       sync.RWMutex
+	vals     map[Key[T]]T
+	handlers []func(o []Event[T], initialSync bool)
+	id       collectionUID
 }
 
-func NewStaticCollection[T any](vals []T) Collection[T] {
+func NewStaticCollection[T any](vals []T) StaticCollection[T] {
 	res := map[Key[T]]T{}
 	for _, v := range vals {
 		res[GetKey(v)] = v
 	}
-	return &staticList[T]{
-		vals: res,
-		id:   nextUID(),
+	return StaticCollection[T]{
+		staticList: &staticList[T]{
+			vals: res,
+			id:   nextUID(),
+		},
+	}
+}
+
+// DeleteObject deletes an object from the collection.
+func (s *staticList[T]) DeleteObject(k Key[T]) {
+	s.mu.Lock() // Unlocked in runEventLocked
+	old, f := s.vals[k]
+	if f {
+		delete(s.vals, k)
+		s.runEventsLocked([]Event[T]{{
+			Old:   &old,
+			Event: controllers.EventDelete,
+		}})
+	} else {
+		s.mu.Unlock()
+	}
+}
+
+// DeleteObjects deletes all objects matching the provided filter
+func (s StaticCollection[T]) DeleteObjects(filter func(obj T) bool) {
+	s.mu.Lock() // Unlocked in runEventLocked
+	var removed []Event[T]
+	for k, v := range s.vals {
+		if filter(v) {
+			delete(s.vals, k)
+			removed = append(removed, Event[T]{
+				Old:   &v,
+				Event: controllers.EventDelete,
+			})
+		}
+	}
+	if len(removed) > 0 {
+		s.runEventsLocked(removed)
+	} else {
+		s.mu.Unlock()
+	}
```

Comment:
this seems brittle, potentially easy to break in a subsequent PR. The mutex needs to be held only until handlers is cloned. Could we take the clone and unlock outside of runEventsLocked and pass in the cloned handlers?